### PR TITLE
Erlang 19.0

### PR DESCRIPTION
**Pull request by erichkist**

### CODE REVIEW

**Comment by JakubCzarniecki** [comment]
> Major releases of erlang should typically start out as separate formula in
Homebrew-versions, and only be merged to master when things like couchdb and
elixir are compatible.

Are we sure we won't break all the things❓

**Comment by erichkist** [comment]
@JakubCzarniecki I install elixir 1.3.0 and couchdb and it seems good. Is there any other test that you suggest?

```bash
~ 2.3.0
$ brew install elixir
==> Downloading https://github.com/elixir-lang/elixir/archive/v1.3.0.tar.gz
Already downloaded: /opt/boxen/cache/homebrew/elixir-1.3.0.tar.gz
==> make
🍺  /opt/boxen/homebrew/Cellar/elixir/1.3.0: 383 files, 5M, built in 1 minute 26 seconds

~ 2.3.0
$ brew install couchdb
==> Installing dependencies for couchdb: nspr, spidermonkey, icu4c
==> Installing couchdb dependency: nspr
==> Downloading https://homebrew.bintray.com/bottles/nspr-4.12.el_capitan.bottle.tar.gz
######################################################################## 100.0%
==> Pouring nspr-4.12.el_capitan.bottle.tar.gz
==> Caveats
This formula is keg-only, which means it was not symlinked into /opt/boxen/homebrew.

Having this library symlinked makes Firefox pick it up instead of built-in,
so it then randomly crashes without meaningful explanation.

Please see https://bugzilla.mozilla.org/show_bug.cgi?id=1142646 for details.

Generally there are no consequences of this for you. If you build your
own software and it requires this formula, you'll need to add to your
build variables:

    LDFLAGS:  -L/opt/boxen/homebrew/opt/nspr/lib
    CPPFLAGS: -I/opt/boxen/homebrew/opt/nspr/include

==> Summary
🍺  /opt/boxen/homebrew/Cellar/nspr/4.12: 91 files, 1.1M
==> Installing couchdb dependency: spidermonkey
==> Downloading https://homebrew.bintray.com/bottles/spidermonkey-1.8.5_1.el_capitan.bottle.1.tar.gz
######################################################################## 100.0%
==> Pouring spidermonkey-1.8.5_1.el_capitan.bottle.1.tar.gz
🍺  /opt/boxen/homebrew/Cellar/spidermonkey/1.8.5_1: 103 files, 10.6M
==> Installing couchdb dependency: icu4c
==> Downloading https://homebrew.bintray.com/bottles/icu4c-57.1.el_capitan.bottle.tar.gz
######################################################################## 100.0%
==> Pouring icu4c-57.1.el_capitan.bottle.tar.gz
==> Caveats
This formula is keg-only, which means it was not symlinked into /opt/boxen/homebrew.

OS X provides libicucore.dylib (but nothing else).

Generally there are no consequences of this for you. If you build your
own software and it requires this formula, you'll need to add to your
build variables:

    LDFLAGS:  -L/opt/boxen/homebrew/opt/icu4c/lib
    CPPFLAGS: -I/opt/boxen/homebrew/opt/icu4c/include

==> Summary
🍺  /opt/boxen/homebrew/Cellar/icu4c/57.1: 265 files, 65.0M
==> Installing couchdb
==> Downloading https://homebrew.bintray.com/bottles/couchdb-1.6.1_5.el_capitan.bottle.tar.gz
######################################################################## 100.0%
==> Pouring couchdb-1.6.1_5.el_capitan.bottle.tar.gz
==> Caveats
To test CouchDB run:
    curl http://127.0.0.1:5984/

The reply should look like:
    {"couchdb":"Welcome","uuid":"....","version":"1.6.1","vendor":{"version":"1.6.1-1","name":"Homebrew"}}

To have launchd start couchdb now and restart at login:
  brew services start couchdb
Or, if you don't want/need a background service you can just run:
  couchdb
==> Summary
🍺  /opt/boxen/homebrew/Cellar/couchdb/1.6.1_5: 638 files, 15.0M

~ 2.3.0
$ erl --version
Erlang/OTP 19 [erts-8.0] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V8.0  (abort with ^G)
1>

~ 2.3.0
$ brew services start couchdb
==> Successfully started `couchdb` (label: homebrew.mxcl.couchdb)

~ 2.3.0
$ curl http://127.0.0.1:5984/
{"couchdb":"Welcome","uuid":"11cc9e05f09e0a2c3794efc58f9aa27e","version":"1.6.1","vendor":{"name":"Homebrew","version":"1.6.1_5"}}

~ 2.3.0
$ iex
Erlang/OTP 19 [erts-8.0] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Interactive Elixir (1.3.0) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)>

```

**Comment by JakubCzarniecki** [comment]
Looking good, thank you. Happy Erlang/OTP 19 release day, @erichkist 🍻

**Comment by erichkist** [comment]
@JakubCzarniecki 💛 💚 💙 ❤️

File: Formula/erlang.rb [hunk 1]
```diff
@@ -40,13 +40,13 @@ class Erlang < Formula
   fails_with :llvm
 
   resource "man" do
-    url "https://www.erlang.org/download/otp_doc_man_18.3.tar.gz"
-    sha256 "978be100e9016874921b3ad1a65ee46b7b6a1e597b8db2ec4b5ef436d4c9ecc2"
+    url "http://erlang.org/download/otp_doc_man_19.0.tar.gz"
```
Comment (DomT4) [inline Formula/erlang.rb]: Seems `https://www.erlang.org/download/otp_doc_man_19.0.tar.gz` still works. We'd like to stick to using `https://` where possible.

File: Formula/erlang.rb [hunk 2]
```diff
@@ -40,13 +40,13 @@ class Erlang < Formula
...
+    sha256 "c7a3d6d85a5a2b96d844297a3fa1bee448c3dd86237734688466249fd5a1401e"
   end
 
   resource "html" do
-    url "https://www.erlang.org/download/otp_doc_html_18.3.tar.gz"
-    sha256 "8fd6980fd05367735779a487df107ace7c53733f52fbe56de7ca7844a355676f"
+    url "http://erlang.org/download/otp_doc_html_19.0.tar.gz"
```
Comment (DomT4) [inline Formula/erlang.rb]: Same comment as above.

File: Formula/erlang.rb (hunk 1 above)
Comment (DomT4) [inline Formula/erlang.rb]: Although the `https://` version generates a different checksum, apparently. Might want to ask upstream about the difference between the two.

Comment (tmjoen) [inline Formula/erlang.rb]: It doesn't look like they [interact much on that twitter account](https://twitter.com/erlang_org/with_replies). Maybe there is some alternate way of communicating with them? IRC? GitHub?

Comment (jeffweiss) [inline Formula/erlang.rb]: it looks like whatever had the inconsistent SHAs is gone. https://www.erlang.org/... has an HTTP redirect to http://erlang.org/... now, and the SHAs are the same.

Comment (erichkist) [inline Formula/erlang.rb]: @tmjoen Yes, you are right. I believe that we no longer need to contact them.

I was checking the `https` and it redirects to `http` version:

```
$ curl -I https://www.erlang.org/download/otp_doc_man_19.0.tar.gz
HTTP/1.1 301 Moved Permanently
Server: nginx
Date: Wed, 22 Jun 2016 21:33:34 GMT
Content-Type: text/html
Content-Length: 184
Connection: keep-alive
Location: http://erlang.org/download/otp_doc_man_19.0.tar.gz
```

In chrome, my https file checksum is different from http file. But in another browsers and curl I don't have this problem.

I delete my Erlang version and update the formula to use the `https` and it works.

@tmjoen, @DomT4 Could you checkout the branch and test it too?

```bash
~/src/personal/homebrew-core 2.3.0 (ek-erlang-19 * u=)
$ g d
diff --git a/Formula/erlang.rb b/Formula/erlang.rb
index d20c281..f1a2d79 100644
--- a/Formula/erlang.rb
+++ b/Formula/erlang.rb
@@ -40,12 +40,12 @@ class Erlang < Formula
   fails_with :llvm

   resource "man" do
-    url "http://erlang.org/download/otp_doc_man_19.0.tar.gz"
+    url "https://www.erlang.org/download/otp_doc_man_19.0.tar.gz"
     sha256 "c7a3d6d85a5a2b96d844297a3fa1bee448c3dd86237734688466249fd5a1401e"
   end

   resource "html" do
-    url "http://erlang.org/download/otp_doc_html_19.0.tar.gz"
+    url "https://www.erlang.org/download/otp_doc_html_19.0.tar.gz"
     sha256 "b6f7c4e964673333f6c3eea8530dd988b41339b8912ae293f6f1b429489159ff"
   end


~/src/personal/homebrew-core 2.3.0 (ek-erlang-19 * u=)
$ brew install Formula/erlang.rb
==> Downloading https://homebrew.bintray.com/bottles/erlang-19.0.el_capitan.bottle.2.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
Error: Failed to download resource "erlang"
Download failed: https://homebrew.bintray.com/bottles/erlang-19.0.el_capitan.bottle.2.tar.gz
Warning: Bottle installation failed: building from source.
==> Downloading https://github.com/erlang/otp/archive/OTP-19.0.tar.gz
==> Downloading from https://codeload.github.com/erlang/otp/tar.gz/OTP-19.0
######################################################################## 100.0%
==> ./otp_build autoconf
==> ./configure --disable-silent-rules --prefix=/opt/boxen/homebrew/Cellar/erlang/19.0 --enable-kernel-poll --enable-threads --enable-sctp --enable-dynamic-ss
==> make
==> make install
==> Downloading https://www.erlang.org/download/otp_doc_man_19.0.tar.gz
==> Downloading from http://erlang.org/download/otp_doc_man_19.0.tar.gz
######################################################################## 100.0%
==> Downloading https://www.erlang.org/download/otp_doc_html_19.0.tar.gz
==> Downloading from http://erlang.org/download/otp_doc_html_19.0.tar.gz
######################################################################## 100.0%
==> Caveats
Man pages can be found in:
  /opt/boxen/homebrew/opt/erlang/lib/erlang/man

Access them with `erl -man`, or add this directory to MANPATH.
==> Summary
🍺  /opt/boxen/homebrew/Cellar/erlang/19.0: 7,292 files, 278.6M, built in 12 minutes 40 seconds
```

Comment (erichkist) [inline Formula/erlang.rb]: @jeffweiss Yes. Sorry, I saw your message after send mine. :)

**Comment by tdsmith** [comment]
@erichkist Can you try installing those packages again using `--build-from-source`? You're installing the bottles, which are precompiled, which doesn't test whether the new version of Erlang can build those packages.

**Comment by erichkist** [comment]
@tdsmith Sure! I hadn't seen it in the first try. Yes, we have a problem with couchdb.

```bash
~/src/personal/homebrew-core 2.3.0 (ek-erlang-19 * u=)
$ brew install elixir couchdb --build-from-source
==> Downloading https://github.com/elixir-lang/elixir/archive/v1.3.0.tar.gz
==> Downloading from https://codeload.github.com/elixir-lang/elixir/tar.gz/v1.3.0
######################################################################## 100.0%
==> make
🍺  /opt/boxen/homebrew/Cellar/elixir/1.3.0: 383 files, 5M, built in 1 minute 22 seconds
==> Downloading https://www.apache.org/dyn/closer.cgi?path=/couchdb/source/1.6.1/apache-couchdb-1.6.1.tar.gz
==> Best Mirror http://mirror.nbtelecom.com.br/apache/couchdb/source/1.6.1/apache-couchdb-1.6.1.tar.gz
######################################################################## 100.0%
==> Patching
patching file INSTALL.Unix
patching file INSTALL.Windows
patching file configure.ac
patching file share/doc/src/install/unix.rst
patching file share/doc/src/install/windows.rst
patching file configure
==> ./configure --prefix=/opt/boxen/homebrew/Cellar/couchdb/1.6.1_5 --localstatedir=/opt/boxen/homebrew/var --sysconfdir=/opt/boxen/homebrew/etc --disable-ini
Last 15 lines from /Users/erichkist/Library/Logs/Homebrew/couchdb/01.configure:
checking whether JSOPTION_ANONFUNFIX is declared... yes
checking for JS_ThrowStopIteration in -lmozjs185... yes
checking for JS_GetStringCharsAndLength in -lmozjs185... yes
checking for JSScript*... no
./configure: line 18138: fg: no job control
checking for icu-config... /opt/boxen/homebrew/opt/icu4c/bin/icu-config
checking for ICU >= 3.4.1... yes
checking ICU_CPPFLAGS... -I/opt/boxen/homebrew/Cellar/icu4c/57.1/include
checking ICU_CFLAGS...
checking ICU_CXXFLAGS... --std=c++0x
checking ICU_LIBS...  -L/opt/boxen/homebrew/Cellar/icu4c/57.1/lib -licui18n -licuuc -licudata
checking for curl-config... /usr/bin/curl-config
checking for curl (libcurl 7.43.0) >= 7.18.0... yes
checking Erlang version compatibility... -n detected Erlang version: 8.0.0...
configure: error: The installed Erlang version must be >= R14B (erts-5.8.1) and =< 18 (erts-7.0) major_version does not match

READ THIS: https://git.io/brew-troubleshooting
If reporting this issue please do so at (not Homebrew/brew):
  https://github.com/Homebrew/homebrew-core/issues

These open issues may also help:
Fold geocouch in to couchdb formula https://github.com/Homebrew/homebrew-core/pull/475
```

I sent a message to the couchdb dev mailing list asking about the dependencies update.

**Comment by KronicDeth** [comment]
I don't think it should block merging this, but in case other people hit the bug after upgrading to Erlang 19.0, I found a dialyzer crash (http://bugs.erlang.org/browse/ERL-177) when using Elixir 1.3 and Erlang 19.0 that does not happen with Elixir 1.3 and Erlang 18.3.

**Comment by JakubCzarniecki** [comment]
As soon as Homebrew/homebrew-versions/pull/1297 is merged we could just use erlang-r18 recipe before issues with 19.0 are resolved.

**Comment by cirla** [comment]
The CouchDB formula [is already patching Erlang/OTP 18.0 compatibility](https://github.com/Homebrew/homebrew-core/blob/447680c2d2030173e89bf5a54830e5d3e4c33214/Formula/couchdb.rb#L137-L266)  ([upstream](https://github.com/apache/couchdb/commit/95cb436)) into 1.6.1.

I don't see a corresponding PR on the upstream [1.6.2](https://github.com/apache/couchdb/tree/1.6.x) or [1.7](https://github.com/apache/couchdb/tree/1.x.x) branches, but something similar may be a shorter-term solution, assuming a similarly patched 1.6.1 builds and functions correctly on Erlang/OTP 19.0.

**Comment by erichkist** [comment]
@tdsmith erlang-r18 was merged: https://github.com/Homebrew/homebrew-versions/pull/1297 (thanks @JakubCzarniecki). Could we merge this one?

**Comment by asabil** [comment]
19.0.1 has been released https://github.com/erlang/otp/releases/tag/OTP-19.0.1

**Comment by DomT4** [comment]
If we can re-patch `couchdb` for the new version & get that submitted upstream, _providing it works_, and roll in the 19.0.1 update, we should be able to move ahead here.

**Comment by erichkist** [comment]
@DomT4 Erlang 19.0.1 doesn't have the docs yet. How can we proceed?

Regarding couchdb, who wants to do that?

**Comment by asabil** [comment]
It's the same docs as 19.0

**Comment by erichkist** [comment]
@asabil Ok. Thanks.

```bash
==> Downloading https://homebrew.bintray.com/bottles/erlang-19.0.1.el_capitan.bottle.2.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
Error: Failed to download resource "erlang"
Download failed: https://homebrew.bintray.com/bottles/erlang-19.0.1.el_capitan.bottle.2.tar.gz
Warning: Bottle installation failed: building from source.
==> Downloading https://github.com/erlang/otp/archive/OTP-19.0.1.tar.gz
Already downloaded: /opt/boxen/cache/homebrew/erlang-19.0.1.tar.gz
==> ./otp_build autoconf
==> ./configure --disable-silent-rules --prefix=/opt/boxen/homebrew/Cellar/erlang/19.0.1 --enable-kernel-poll --enable-threads --enable-sctp --enable-dynami
==> make
==> make install
==> Downloading https://www.erlang.org/download/otp_doc_man_19.0.tar.gz
==> Downloading from http://erlang.org/download/otp_doc_man_19.0.tar.gz
######################################################################## 100.0%
==> Downloading https://www.erlang.org/download/otp_doc_html_19.0.tar.gz
==> Downloading from http://erlang.org/download/otp_doc_html_19.0.tar.gz
######################################################################## 100.0%
==> Caveats
Man pages can be found in:
  /opt/boxen/homebrew/opt/erlang/lib/erlang/man

Access them with `erl -man`, or add this directory to MANPATH.
==> Summary
🍺  /opt/boxen/homebrew/Cellar/erlang/19.0.1: 7,292 files, 278.6M, built in 13 minutes 4 seconds
```

**Comment by MikeMcQuaid** [comment]
> Regarding couchdb, who wants to do that?

@erichkist Could you give it a go?

**Comment by gmile** [comment]
@MikaelSmith I've edited the couchdb formula (the patch section) to support R19.0 – https://github.com/Homebrew/homebrew-core/pull/2891, but I'm not sure if it's legitimate to do in such a manner, e.g. forging a contents of commit that belongs to another developer. My gut tells me it shouldn't be legit?

**Comment by DomT4** [comment]
Here's hoping this all goes peacefully.

Merged in https://github.com/Homebrew/homebrew-core/commit/1ffef3a015c00c990420e59fe36068f9b67b22c2 & https://github.com/Homebrew/homebrew-core/commit/0232c3f3648ceb53787d7ca2c8b45ec38a9b1684. Thank you for your contributions to Homebrew @erichkist, and thank you to everyone for the patience & willingness to work on getting this wrapped. Appreciated.

**Pull request closed without merging**